In [1]:
from astropy.table import QTable
import astropy.units as u
import numpy as np
from astropy.table import Table
from astroquery.lamda import Lamda
from astroquery.splatalogue import Splatalogue
from astropy.io import ascii


In [35]:
(2.319009278000E+11*u.Hz).to(u.GHz)

<Quantity 231.9009278 GHz>

In [13]:
Lamda.molecule_dict

{'catom': 'http://home.strw.leidenuniv.nl/~moldata/datafiles/catom.dat',
 'c+': 'http://home.strw.leidenuniv.nl/~moldata/datafiles/c+.dat',
 'c+@uv': 'http://home.strw.leidenuniv.nl/~moldata/datafiles/c+@uv.dat',
 'oatom@lique': 'http://home.strw.leidenuniv.nl/~moldata/datafiles/oatom@lique.dat',
 'n+': 'http://home.strw.leidenuniv.nl/~moldata/datafiles/n+.dat',
 'arh+': 'http://home.strw.leidenuniv.nl/~moldata/datafiles/arh+.dat',
 'cf+': 'http://home.strw.leidenuniv.nl/~moldata/datafiles/cf+.dat',
 'ch-nohfs': 'http://home.strw.leidenuniv.nl/~moldata/datafiles/ch-nohfs.dat',
 'ch-h2': 'http://home.strw.leidenuniv.nl/~moldata/datafiles/ch-h2.dat',
 'ch-h': 'http://home.strw.leidenuniv.nl/~moldata/datafiles/ch-h.dat',
 'cn': 'http://home.strw.leidenuniv.nl/~moldata/datafiles/cn.dat',
 'cn-hfs': 'http://home.strw.leidenuniv.nl/~moldata/datafiles/cn-hfs.dat',
 'co': 'http://home.strw.leidenuniv.nl/~moldata/datafiles/co.dat',
 '13co': 'http://home.strw.leidenuniv.nl/~moldata/datafiles/13c

In [2]:
h2co_321_220_A = 10**Splatalogue.query_lines(218.5*u.GHz, 218.8*u.GHz, chemical_name=" H2CO ")[2]['Log<sub>10</sub> (A<sub>ij</sub>)']

In [31]:
h2co_321_220_freq = Splatalogue.query_lines(218.5*u.GHz, 218.8*u.GHz, chemical_name=" H2CO ")[2]['Freq-GHz(rest frame,redshifted)']

In [28]:
h2co_322_221_A = 10**Splatalogue.query_lines(218.3*u.GHz, 218.5*u.GHz, chemical_name=" H2CO ")[2]['Log<sub>10</sub> (A<sub>ij</sub>)']

In [29]:
h2co_322_221_freq = Splatalogue.query_lines(218.3*u.GHz, 218.5*u.GHz, chemical_name=" H2CO ")[2]['Freq-GHz(rest frame,redshifted)']

In [4]:
collrates, radtransitions, enlevels = Lamda.query(mol='co')
collrates_h2co, radtransitions_h2co, enlevels_h2co = Lamda.query(mol='ph2co-h2')
collrates_ocs, radtransitions_ocs, enlevels_ocs = Lamda.query(mol='ocs@xpol')
collrates_hc3n, radtransitions_hc3n, enlevels_hc3n = Lamda.query(mol='hc3n-h2')
collrates_ch3oh, radtransitions_ch3oh, enlevels_ch3oh = Lamda.query(mol='ch3oh_a')


/blue/adamginsburg/adamginsburg/repos/astroquery/astroquery/lamda/core.py:147: UserWarning: The first time a LAMDA function is called, it must assemble a list of valid molecules and URLs.  This list will be cached so future operations will be faster.
  warnings.warn("The first time a LAMDA function is called, it must "


In [5]:
collrates_ch3oh, radtransitions_ch3oh, enlevels_ch3oh = Lamda.query(mol='e-ch3oh')


In [6]:
collrates_sio, radtransitions_sio, enlevels_sio = Lamda.query(mol='sio-h2')


In [7]:
collrates_13co, radtransitions_13co, enlevels_13co = Lamda.query(mol='13co')
collrates_c18o, radtransitions_c18o, enlevels_c18o = Lamda.query(mol='c18o')


In [8]:
radtransitions_c18o[1]['EinsteinA']

6.011e-07

In [9]:
collrates_c18o['PH2'][2]['C_ij(T=60)']

6e-11

In [10]:
collrates_ch3oh['H2'][156]['C_ij(T=60)']

9.3e-13

In [11]:
collrates_hc3n['PH2'][299]['C_ij(T=60)']#['H2'][170]['C_ij(T=60)']#['PH2']#[5]['C_ij(T=60)']#[2]['C_ij(T=60)']# transition 24-23 aka 25-24

4.71e-11

In [12]:
radtransitions_hc3n[23]['EinsteinA']#[1]['EinsteinA']

0.000826

In [21]:
radtransitions[1]['Frequency']

230.538

In [13]:
#t.add_row(['\ce{OCS} $J=18\shortrightarrow17$', 218.90*u.GHz, 960, 321.2*u.km/u.second, 0.774*u.km/u.second])
#t.add_row(['\ce{CH3OH} $J=4_{2,2}\shortrightarrow3_{1,2}$', 218.44*u.GHz, 960, 321.9*u.km/u.second, 0.671*u.km/u.second])
#t.add_row(['\ce{p-H2CO} $J=3_{22}\shortrightarrow2_{21}$', 218.48*u.GHz, 960, 321.9*u.km/u.second, 0.671*u.km/u.second])

In [36]:
names = ('Molecule and Transition',
         'Center Rest Frequency', 
         'Einstein A',
         'Collision Rates (T = \SI{60}{K})',
         'Critical Density',
        'Eff. Ch. #', 
        'Velocity Bandwidth',
        'Velocity Resolution')

spw = [ '\ce{^12CO} $J=2\shortrightarrow1$',
        'H(30)$\alpha$',
        '\ce{H2CO} $J=3_{2,1}\shortrightarrow2_{2,0}$',
         '\ce{OCS} $J=18\shortrightarrow17$', # OCS
        '\ce{HC3N} v=0 $J=24\shortrightarrow23$',
         '\ce{CH3OH} $J=4_{2,2}\shortrightarrow3_{1,2}$', # CH3OH
         '\ce{H2CO} $J=3_{2,2}\shortrightarrow2_{2,1}$', # extra H2CO
        '\ce{H2CO} $J=3_{0,3}\shortrightarrow2_{0,2}$',
        'SiO v=0 $J=5\shortrightarrow4$',
        '\ce{^13CO} $J=2\shortrightarrow1$',
        '\ce{C^18O} $J=2\shortrightarrow1$']

freq = [radtransitions[1]['Frequency'],
             (2.319009278000E+11*u.Hz).to(u.GHz).value,
             h2co_321_220_freq,
             radtransitions_ocs[17]['Frequency'],
             radtransitions_hc3n[23]['Frequency'],
             radtransitions_ch3oh[239]['Frequency'],
             h2co_322_221_freq,
             radtransitions_h2co[2]['Frequency'],
             radtransitions_sio[4]['Frequency'],
             radtransitions_13co[1]['Frequency'],
             radtransitions_c18o[1]['Frequency']
             ]*u.GHz

       #[230.54,    # 12 CO
       #231.90,  # H30
       #218.76,   # H2CO 321-220
       # 218.90, # OCS
       #218.32,  # HC3N
       # 218.44, # CH3OH
       # 218.48, # extra H2CO
       #218.22,  # H2CO 303-202
       #217.10,  # SiO 5-4
       #220.40,  # 13CO
       #219.56  # C18O
       #] * u.GHz

Einstein_A = [radtransitions[1]['EinsteinA'],
             1,
             h2co_321_220_A,
             radtransitions_ocs[17]['EinsteinA'],
             radtransitions_hc3n[23]['EinsteinA'],
             radtransitions_ch3oh[239]['EinsteinA'],
             h2co_322_221_A,
             radtransitions_h2co[2]['EinsteinA'],
             radtransitions_sio[4]['EinsteinA'],
             radtransitions_13co[1]['EinsteinA'],
             radtransitions_c18o[1]['EinsteinA']
             ]*u.second**-1

collision_rates = [collrates['PH2'][2]['C_ij(T=60)'],
                  1,
                  collrates_h2co['PH2'][5]['C_ij(T=60)'],
                  collrates_ocs['H2'][170]['C_ij(T=60)'],
                  collrates_hc3n['PH2'][299]['C_ij(T=60)'],
                  collrates_ch3oh['H2'][156]['C_ij(T=60)'],
                  collrates_h2co['PH2'][5]['C_ij(T=60)'],
                  collrates_h2co['PH2'][5]['C_ij(T=60)'],
                  collrates_sio['PH2'][14]['C_ij(T=60)'],
                  collrates_13co['PH2'][2]['C_ij(T=60)'],
                  collrates_c18o['PH2'][2]['C_ij(T=60)']] * u.cm**3 * u.second**-1

ncrit = Einstein_A / collision_rates
        #[radtransitions[1]['EinsteinA']/collrates['PH2'][2]['C_ij(T=60)'], # co 2-1
        #0,
        #h2co_321_220_A/collrates_h2co['PH2'][5]['C_ij(T=60)'], # H2CO 321-220
        #radtransitions_ocs[17]['EinsteinA']/collrates_ocs['H2'][170]['C_ij(T=60)'], # OCS 18-17
        #radtransitions_hc3n[23]['EinsteinA']/collrates_hc3n['PH2'][299]['C_ij(T=60)'], # 24-23
        #radtransitions_ch3oh[239]['EinsteinA']/collrates_ch3oh['H2'][156]['C_ij(T=60)'], # 422-312
        #h2co_322_221_A/collrates_h2co['PH2'][5]['C_ij(T=60)'], # H2CO 322 - 221
        #radtransitions_h2co[2]['EinsteinA']/collrates_h2co['PH2'][5]['C_ij(T=60)'], # H2CO 303-202
        #radtransitions_sio[4]['EinsteinA']/collrates_sio['PH2'][14]['C_ij(T=60)'], # 5-4
        #radtransitions_13co[1]['EinsteinA']/collrates_13co['PH2'][2]['C_ij(T=60)'], #13CO 2-1
        #radtransitions_c18o[1]['EinsteinA']/collrates_c18o['PH2'][2]['C_ij(T=60)'] # C18O 2-1
        #]*u.cm**(-3)

ch = [3840, #12 CO  
     3840,  #H30   
     960,   # H2CO 322-221   
      960,
     960,   # HC3N   
      960,
      960,
     960,   # H2CO 303-202     
     960,   # SiO 5-4 
     1920,  # 13CO  
     1920]  # C18O  

band = [304.8, #12 CO
       2424.2, #H30
       321.2,  # H2CO 322-221  
        321.2,
       321.9,  # HC3N  
        321.9,
        321.9,
       322.0,  # H2CO 303-202  
       323.7,  # SiO 5-4  
       318.8,  # 13CO  
       320.1   #C18O
       ]*u.km/u.second

reso = [0.159, #12 CO 
       1.263,  #H30   
       0.774,  # H2CO 322-221   
        0.774,
       0.671,  # HC3N    
        0.671,
        0.671,
       0.671,  # H2CO 303-202   
       0.674,  # SiO 5-4  
       0.332,  # 13CO  
       0.333   # C18O  
        ] *u.km/u.second




In [37]:
np.around(Einstein_A.to(10**-6 * u.second**-1), 2)

<Quantity [6.9000e-01, 1.0000e+06, 2.5481e+02, 3.0370e+01, 8.2600e+02,
           4.6860e+01, 2.5382e+02, 2.8180e+02, 5.1970e+02, 6.0000e-01,
           6.0000e-01] 1e-06 / s>

In [38]:
t = Table([spw, freq, 
           np.around(Einstein_A.to(10**-6 * u.second**-1), 3), 
           np.around(collision_rates.to(10**-11 * u.cm**3 * u.second**-1), 3), 
           np.around(ncrit.to(10**5 * u.cm**-3), 3), 
           ch, band, reso],names=names)
t

<Table length=11>
           Molecule and Transition            Center Rest Frequency Einstein A ... Eff. Ch. # Velocity Bandwidth Velocity Resolution
                                                       GHz          1e-06 / s  ...                  km / s              km / s      
                    str45                            float64         float64   ...   int64         float64             float64      
--------------------------------------------- --------------------- ---------- ... ---------- ------------------ -------------------
            \ce{^12CO} $J=2\shortrightarrow1$               230.538      0.691 ...       3840              304.8               0.159
                                 H(30)$lpha$           231.9009278  1000000.0 ...       3840             2424.2               1.263
 \ce{H2CO} $J=3_{2,1}\shortrightarrow2_{2,0}$            218.760071    254.812 ...        960              321.2               0.774
            \ce{OCS} $J=18\shortrightarrow17$           218.9033565     30.371 ...        960              321.2               0.774
       \ce{HC3N} v=0 $J=24\shortrightarrow23$             218.32472      826.0 ...        960              321.9               0.671
\ce{CH3OH} $J=4_{2,2}\shortrightarrow3_{1,2}$            218.440063     46.863 ...        960              321.9               0.671
 \ce{H2CO} $J=3_{2,2}\shortrightarrow2_{2,1}$            218.475642    253.822 ...        960              321.9               0.671
 \ce{H2CO} $J=3_{0,3}\shortrightarrow2_{0,2}$            218.222192      281.8 ...        960              322.0               0.671
               SiO v=0 $J=5\shortrightarrow4$            217.104919      519.7 ...        960              323.7               0.674
            \ce{^13CO} $J=2\shortrightarrow1$        220.3986841281      0.604 ...       1920              318.8               0.332
            \ce{C^18O} $J=2\shortrightarrow1$           219.5603541      0.601 ...       1920              320.1               0.333

In [39]:
ascii.write(t, format='latex')

\begin{table}
\begin{tabular}{cccccccc}
Molecule and Transition & Center Rest Frequency & Einstein A & Collision Rates (T = \SI{60}{K}) & Critical Density & Eff. Ch. # & Velocity Bandwidth & Velocity Resolution \\
 & $\mathrm{GHz}$ & $\mathrm{1 \times 10^{-6}\,s^{-1}}$ & $\mathrm{1 \times 10^{-11}\,cm^{3}\,s^{-1}}$ & $\mathrm{100000\,cm^{-3}}$ &  & $\mathrm{km\,s^{-1}}$ & $\mathrm{km\,s^{-1}}$ \\
\ce{^12CO} $J=2\shortrightarrow1$ & 230.538 & 0.691 & 6.0 & 0.115 & 3840 & 304.8 & 0.159 \\
H(30)$lpha$ & 231.9009278 & 1000000.0 & 100000000000.0 & 0.0 & 3840 & 2424.2 & 1.263 \\
\ce{H2CO} $J=3_{2,1}\shortrightarrow2_{2,0}$ & 218.760071 & 254.812 & 9.1 & 28.001 & 960 & 321.2 & 0.774 \\
\ce{OCS} $J=18\shortrightarrow17$ & 218.9033565 & 30.371 & 7.4 & 4.104 & 960 & 321.2 & 0.774 \\
\ce{HC3N} v=0 $J=24\shortrightarrow23$ & 218.32472 & 826.0 & 4.71 & 175.372 & 960 & 321.9 & 0.671 \\
\ce{CH3OH} $J=4_{2,2}\shortrightarrow3_{1,2}$ & 218.440063 & 46.863 & 0.093 & 503.904 & 960 & 321.9 & 0.671 \\
\ce

In [20]:
len(freq)

8

In [3]:
# 12CO
row1 = [spw[0], freq[0], ch[0], band[0], reso[0]]
row2 = [spw[1], freq[1], ch[1], band[1], reso[1]]

In [5]:
#a = np.array([1, 4, 5], dtype=np.int32)
#b = [2.0, 5.0, 8.5]
#c = ['x', 'y', 'z']
#d = [10, 20, 30] * u.m / u.s

t = QTable(spw, [freq, ch, band, reso],
           names=names,
           meta={'name': 'first table'})

ValueError: masked should be one of True, False, None

In [32]:
# OCS
t.add_row(['\ce{OCS} $J=18\shortrightarrow17$', 218.90*u.GHz, 960, 321.2*u.km/u.second, 0.774*u.km/u.second])

ValueError: Unable to insert row because of exception in column 'Molecule and Transition':
Cannot parse "\ce{OCS} $J=18\shortrightarrow17$" as a Quantity. It does not start with a number.

In [24]:
# HC3N
t.add_row([freq[3], spw[3], ch[3], band[3], reso[3]])

In [25]:
# CH3OH
t.add_row(['\ce{CH3OH} $J=4_{2,2}\shortrightarrow3_{1,2}$', 218.44*u.GHz, 960, 321.9*u.km/u.second, 0.671*u.km/u.second])

In [26]:
# extra H2CO
t.add_row(['\ce{p-H2CO} $J=3_{22}\shortrightarrow2_{21}$', 218.48*u.GHz, 960, 321.9*u.km/u.second, 0.671*u.km/u.second])

In [27]:
# H2CO 303-202
t.add_row([freq[4], spw[4], ch[4], band[4], reso[4]])
# SiO 5-4
t.add_row([freq[5], spw[5], ch[5], band[5], reso[5]])
# 13CO
t.add_row([freq[6], spw[6], ch[6], band[6], reso[6]])
# C18O
t.add_row([freq[7], spw[7], ch[7], band[7], reso[7]])


In [28]:
from astropy.io import ascii


In [127]:
ascii.write(t, format='latex')

AttributeError: 'builtin_function_or_method' object has no attribute 'write'

\begin{table*}[ht!]
\centering
\caption{Correlator Configuration - The observation parameters of each spectral window. 
 \label{test table}}

\begin{tabular}{|l|l|l|l|l|l|} 
\hline
\label{tab:obs}
\centering
\begin{tabular}[c]{@{}l@{}}Center Rest \\Freq (GHz)\end{tabular} & Spectral Window Name           & Eff. Ch \# & \begin{tabular}[c]{@{}l@{}}Vel. Bandwidth\\(km/s)\end{tabular} & \begin{tabular}[c]{@{}l@{}}Vel. Resolution\\(km/s)\end{tabular}  \\ 
\hline
230.538000                                                       & CO 2-1         & 3840       & 304.8                                                          & 0.159                                                            \\ 
\hline
231.900928                                                       & H(30)              & 3840       & 2424.2                                                         & 1.263                                                            \\ 
\hline
218.760066                                                       & H$_2$CO 3(2,2)-2(2,1) & 960        & 321.2                                                          & 0.774                                                            \\ 
\hline
218.324723                                                       & HC$_3$N v=0 J=24-23   & 960        & 321.9                                                          & 0.671                                                            \\ 
\hline
218.222192                                                       & H$_2$CO 3(0,3)-2(0,2) & 960        & 322.0                                                          & 0.671                                                            \\ 
\hline
217.104980                                                       & SiO v=0 5-4        & 960        & 323.7                                                          & 0.674                                                            \\ 
\hline
220.398684                                                       & 13CO 2-1       & 1920       & 318.8                                                          & 0.332                                                            \\ 
\hline
219.560358                                                       & C18O 2-1           & 1920       & 320.1                                                          & 0.333                                                            \\
\hline
\end{tabular}

\end{table*}